ui/components/component-library/button-base/button-base.tsx

```diff
@@ -0,0 +1,126 @@
+import React from 'react';
+import classnames from 'classnames';
+import { IconName, Icon, IconSize, Text } from '..';
+import {
+  AlignItems,
+  Display,
+  JustifyContent,
+  TextColor,
+  TextVariant,
+  BorderRadius,
+  BackgroundColor,
+  IconColor,
+} from '../../../helpers/constants/design-system';
+import type { PolymorphicRef } from '../box';
+import {
+  ButtonBaseProps,
+  ButtonBaseSize,
+  ButtonBaseComponent,
+} from './button-base.types';
+
+export const ButtonBase: ButtonBaseComponent = React.forwardRef(
+  <C extends React.ElementType = 'button' | 'a'>(
+    {
+      block,
+      children,
+      className = '',
+      href,
+      ellipsis = false,
+      externalLink,
+      size = ButtonBaseSize.Md,
+      startIconName,
+      startIconProps,
+      endIconName,
+      endIconProps,
+      loading,
+      disabled,
```

Comment:
There are a few props here that I feel need to be conditionally passed based on the element the `as` prop is receiving. e.g. `disabled`, `ref`, `href` etc are all attributes for either the `button` or `a` component but are mutually exclusive for each element. I'm quite surprised this is passing CI. But you can see the linter complain

![Screenshot 2023-07-19 at 2 14 58 PM](https://github.com/MetaMask/metamask-extension/assets/8112138/accd9845-6f90-49e9-bf11-81940035cebd)

Another way to describe this is we shouldn't allow link attributes to be attached to a `button` element and vice versa for the `disabled` attribute for link

<img width="1438" alt="Screenshot 2023-07-19 at 2 19 24 PM" src="https://github.com/MetaMask/metamask-extension/assets/8112138/8cbc1d02-8121-4cd0-9b26-706b978aebea">

The way the polymorphic typing works is that it excludes the component props from the HTML element checking which is why it accepts `ref` and `target` from `externalLink`